### PR TITLE
arch/arm: syscall SYS_switch_context and SYS_restore_context use 0 para

### PR DESCRIPTION
**Pull request by hujun260**
## Summary
arch/arm: syscall SYS_switch_context and SYS_restore_context use 0 para
reason:
simplify context switch
sys_call0(SYS_switch_context)
sys_call0(SYS_restore_context)

## Impact
arm

## Testing
ostest
mps2-an500:nsh
qemu-armv7a:smp

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



No, this PR does not fully meet the NuttX requirements.  While it provides some information, it lacks crucial details. Here's a breakdown:

* **Insufficient Summary:** While it states the *what*, it's missing the *why* and *how*.  Why is simplifying the context switch necessary? What specific changes were made to achieve this? How does using 0 parameters simplify it?  What were the parameters before?

* **Incomplete Impact:**  Simply stating "arm" is not enough. Which ARM architectures?  Which boards?  Does this change affect any existing drivers? Does this have performance implications? Does it change any APIs that applications might use? It needs more specifics.  Address all impact categories explicitly (even if the answer is NO with a brief justification).

* **Missing Testing Details:** Listing platforms is insufficient.  "ostest", "mps2-an500:nsh", and "qemu-armv7a:smp" are not descriptive enough.  What specific tests were run? *Crucially, the testing logs are empty.*  The logs demonstrating the before and after behavior are mandatory for demonstrating the change's correctness and verifying its impact.  Provide actual command-line invocations and their output.  Also specify the Build Host details.

**Example of how to improve this PR description:**

## Summary
arch/arm: Simplify context switching by removing parameters from SYS_switch_context and SYS_restore_context

This change removes the parameters previously required for `SYS_switch_context` and `SYS_restore_context` syscalls.  These parameters were redundant as the necessary context information is already managed internally within the scheduler.  This simplification reduces overhead in context switching and improves code clarity.  No related NuttX issue.

## Impact

* **Is new feature added? Is existing feature changed?**  Existing feature changed (optimization).
* **Impact on user:** YES. Applications directly using these syscalls (which is unlikely as these are typically used internally) will need to be updated to remove the parameters.
* **Impact on build:** NO.
* **Impact on hardware:** YES.  This change affects ARM architectures (ARMv7-M, ARMv7-A, Cortex-M4 tested). Tested on `mps2-an500` and simulated `qemu-armv7a:smp`.
* **Impact on documentation:** YES. The syscall documentation needs to be updated to reflect the parameter removal.  A documentation update is included in this PR.
* **Impact on security:** NO.
* **Impact on compatibility:** Backward compatibility is broken for applications directly using these syscalls.  Forward compatibility is maintained.
* **Anything else to consider:** Potential performance improvements.

## Testing

I confirm that changes are verified on local setup and works as intended:

* **Build Host:** Linux (Ubuntu 20.04), x86_64, GCC 9.4.0
* **Target(s):** ARMv7-M (sim:qemu-armv7m), ARMv7-A (sim:qemu-armv7a:smp), Cortex-M4 (mps2-an500:nsh)

Testing logs before change:

```
# On mps2-an500:nsh
nsh> <command to demonstrate previous behavior with parameters>
<output demonstrating previous behavior>
```

Testing logs after change:

```
# On mps2-an500:nsh
nsh> <command to demonstrate new behavior without parameters>
<output demonstrating new behavior>

# On qemu-armv7a:smp
<commands and output demonstrating SMP functionality>

# ostest output (relevant sections) before and after
<ostest output before>
<ostest output after>
```


By providing complete and detailed information, the reviewers can effectively evaluate the PR and ensure its quality and correctness. Remember, the goal is to make the change clear, easy to understand, and demonstrate its impact and correctness.

File: arch/arm/include/irq.h [hunk 1]
```diff
@@ -82,13 +82,13 @@
...
 
 #ifndef up_switch_context
-#define up_switch_context(tcb, rtcb)                              \
-  do {                                                            \
-    if (!up_interrupt_context())                                  \
-      {                                                           \
-        sys_call2(SYS_switch_context, (uintptr_t)&rtcb->xcp.regs, \
-                  (uintptr_t)tcb->xcp.regs);                      \
-      }                                                           \
+#define up_switch_context(tcb, rtcb)   \
```
Comment (xiaoxiang781216) [inline arch/arm/include/irq.h:85]: should we remove tcb and rtcb too?

Comment (hujun260) [reply arch/arm/include/irq.h:85]: When all architectures have been modified.  we could remove tcb and rtcb

File: arch/arm/src/armv8-m/arm_svcall.c [hunk 2]
```diff
@@ -457,12 +414,12 @@ int arm_svcall(int irq, void *context, void *arg)
...
 
-  if (regs != new_regs)
+  if (need_switch)
     {
-      restore_critical_section(tcb, this_cpu());
+      restore_critical_section(this_task(), this_cpu());
 
 #ifdef CONFIG_DEBUG_SYSCALL_INFO
-      regs = new_regs;
+      regs = this_task()->xcp.regs;
```
[the remark this answers is not in the document]
Comment (hujun260) [reply arch/arm/src/armv8-m/arm_svcall.c]: done

File: arch/arm/src/common/arm_internal.h [hunk 3]
```diff
@@ -147,8 +147,8 @@
 /* Context switching */
 
 #ifndef arm_fullcontextrestore
-#  define arm_fullcontextrestore(restoreregs) \
-    sys_call1(SYS_restore_context, (uintptr_t)restoreregs);
+#  define arm_fullcontextrestore() \
+    sys_call0(SYS_restore_context);
 #else
```
[the remark this answers is not in the document]
Comment (hujun260) [reply arch/arm/src/common/arm_internal.h:151]: There is a special implementation:
tc32_fullcontextrestore

File: arch/arm/src/common/arm_internal.h [hunk 4]
```diff
@@ -147,8 +147,8 @@
 /* Context switching */
 
 #ifndef arm_fullcontextrestore
-#  define arm_fullcontextrestore(restoreregs) \
-    sys_call1(SYS_restore_context, (uintptr_t)restoreregs);
+#  define arm_fullcontextrestore() \
+    sys_call0(SYS_restore_context);
```
[the remark this answers is not in the document]
Comment (hujun260) [reply arch/arm/src/common/arm_internal.h]: done

File: arch/arm/src/common/arm_internal.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply arch/arm/src/common/arm_internal.h:151]: should we let's tc32 define arm_fullcontextrestore directly?

[the remark this answers is not in the document]
Comment (hujun260) [reply arch/arm/src/common/arm_internal.h]: done

File: arch/arm/src/armv6-m/arm_svcall.c [hunk 5]
```diff
@@ -437,30 +410,23 @@ int arm_svcall(int irq, void *context, void *arg)
...
-#endif
-    }
 #ifdef CONFIG_DEBUG_SYSCALL_INFO
-  else
+#  ifndef CONFIG_DEBUG_SVCALL
+  if (cmd > SYS_switch_context)
+#  endif
     {
-      svcinfo("SVCall Return: %d\n", regs[REG_R0]);
+  svcinfo("SVCall Return:\n");
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv6-m/arm_svcall.c:418]: wrong indent

File: arch/arm/src/armv7-m/arm_svcall.c [hunk 6]
```diff
@@ -446,30 +419,23 @@ int arm_svcall(int irq, void *context, void *arg)
...
-#endif
-    }
-#ifdef CONFIG_DEBUG_SYSCALL_INFO
-  else
+#  ifndef CONFIG_DEBUG_SVCALL
+  if (cmd > SYS_switch_context)
+#  endif
     {
-      svcinfo("SVCall Return: %d\n", regs[REG_R0]);
+  svcinfo("SVCall Return:\n");
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv7-m/arm_svcall.c:427]: ditto

File: arch/arm/src/armv8-m/arm_svcall.c [hunk 7]
```diff
@@ -446,30 +419,23 @@ int arm_svcall(int irq, void *context, void *arg)
...
-#endif
-    }
-#ifdef CONFIG_DEBUG_SYSCALL_INFO
-  else
+#  ifndef CONFIG_DEBUG_SVCALL
+  if (cmd > SYS_switch_context)
+#  endif
     {
-      svcinfo("SVCall Return: %d\n", regs[REG_R0]);
+  svcinfo("SVCall Return:\n");
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv8-m/arm_svcall.c:427]: ditto

File: arch/arm/include/tlsr82/irq.h [hunk 8]
```diff
@@ -283,6 +283,8 @@ static inline_function void up_set_interrupt_context(bool flag)
 #endif
 }
 
+#define arm_fullcontextrestore() tc32_fullcontextrestore(this_task()->xcp.regs)
```
Comment (xiaoxiang781216) [inline arch/arm/include/tlsr82/irq.h:286]: already define in: arch/arm/src/tlsr82/Make.defs

File: arch/arm/src/armv8-m/arm_svcall.c (hunk 7 above)
Comment (hujun260) [reply arch/arm/src/armv8-m/arm_svcall.c]: done

File: arch/arm/include/tlsr82/irq.h (hunk 8 above)
Comment (xiaoxiang781216) [reply arch/arm/include/tlsr82/irq.h:286]: but should remove this definition.

File: arch/arm/src/armv8-r/arm_syscall.c [hunk 9]
```diff
@@ -254,7 +261,20 @@ uint32_t *arm_syscall(uint32_t *regs)
 #endif
 
       case SYS_restore_context:
+        switch_context = true;
+
+        /* Update scheduler parameters */
+
+        nxsched_resume_scheduler(tcb);
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv8-r/arm_syscall.c:268]: why not call nxsched_suspend_scheduler

File: arch/arm/src/armv8-r/arm_syscall.c [hunk 10]
```diff
@@ -254,7 +261,20 @@ uint32_t *arm_syscall(uint32_t *regs)
...
+        break;
       case SYS_switch_context:
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv8-r/arm_syscall.c:270]: could merge the code block of SYS_restore_context and SYS_switch_context.

File: arch/arm/src/armv8-r/arm_syscall.c [hunk 11]
```diff
@@ -520,19 +540,8 @@ uint32_t *arm_syscall(uint32_t *regs)
...
-
-      /* Record the new "running" task.  g_running_tasks[] is only used by
-       * assertion logic for reporting crashes.
-       */
-
-      *running_task = tcb;
-
       /* Restore the cpu lock */
 
       restore_critical_section(tcb, this_cpu());
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv8-r/arm_syscall.c:547]: move into SYS_switch_context block.

Comment (hujun260) [reply arch/arm/src/armv8-r/arm_syscall.c]: ok

File: arch/arm/src/armv8-r/arm_syscall.c (hunk 10 above)
Comment (hujun260) [reply arch/arm/src/armv8-r/arm_syscall.c]: done

File: arch/arm/src/common/arm_internal.h [hunk 12]
```diff
@@ -147,10 +149,7 @@
 /* Context switching */
 
 #ifndef arm_fullcontextrestore
-#  define arm_fullcontextrestore(restoreregs) \
-    sys_call1(SYS_restore_context, (uintptr_t)restoreregs);
-#else
-extern void arm_fullcontextrestore(uint32_t *restoreregs);
+#  define arm_fullcontextrestore() sys_call0(SYS_restore_context);
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_internal.h:152]: remove ;

Comment (hujun260) [reply arch/arm/src/common/arm_internal.h]: done

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by hartmannathan** [comment]
Please see Issue #15100

**Pull request merged**